### PR TITLE
Make back and forward buttons work in FileManager

### PR DESCRIPTION
**Pull request by christopherdumas**
Also fixed it so that directories don't get double-opened (first when they are opened, and second when the selection changes to match in the file tree view). Fixes: #36.

### CODE REVIEW

File: Applications/FileManager/DirectoryView.cpp [hunk 1]
```diff
@@ -46,6 +46,8 @@ void DirectoryView::handle_activation(const GModelIndex& index)
 DirectoryView::DirectoryView(GWidget* parent)
     : GStackWidget(parent)
     , m_model(GDirectoryModel::create())
+    , path_history_position(0)
```
Comment (awesomekling) [inline Applications/FileManager/DirectoryView.cpp]: Please initialize this in the header instead;
`int path_history_position { 0 };`

File: Applications/FileManager/DirectoryView.cpp [hunk 2]
```diff
@@ -46,6 +46,8 @@ void DirectoryView::handle_activation(const GModelIndex& index)
...
+    , path_history()
```
Comment (awesomekling) [inline Applications/FileManager/DirectoryView.cpp]: You can omit constructor initializers for default-constructed objects.

File: Applications/FileManager/DirectoryView.cpp [hunk 3]
```diff
@@ -108,8 +110,32 @@ void DirectoryView::set_view_mode(ViewMode mode)
     ASSERT_NOT_REACHED();
 }
 
+void print_path_history(int index, Vector<String> path_history)
+{
+    if (path_history.size() > 0) {
```
Comment (awesomekling) [inline Applications/FileManager/DirectoryView.cpp]: Please prefer "early return" style to reduce nesting.
```suggestion
    if (path_history.is_empty())
        return;
```

File: Applications/FileManager/DirectoryView.h [hunk 4]
```diff
@@ -36,6 +39,9 @@ class DirectoryView final : public GStackWidget {
     ViewMode m_view_mode { Invalid };
 
     Retained<GDirectoryModel> m_model;
+    int path_history_position;
```
Comment (rburchell) [inline Applications/FileManager/DirectoryView.h]: style nit: m_ prefix on members - and it might be a good idea to initialize the int in the header too for consistency:

    int m_path_history_position { 0 };

File: Applications/FileManager/DirectoryView.cpp [hunk 5]
```diff
@@ -108,8 +110,32 @@ void DirectoryView::set_view_mode(ViewMode mode)
     ASSERT_NOT_REACHED();
 }
 
+void print_path_history(int index, Vector<String> path_history)
```
Comment (awesomekling) [inline Applications/FileManager/DirectoryView.cpp]: 'path_history' should be passed by const-reference (const Vector<String>&) as this avoids making an unnecessary copy of the whole vector.

File: Applications/FileManager/DirectoryView.cpp (hunk 3 above)
Comment (rburchell) [inline Applications/FileManager/DirectoryView.cpp]: Could just print this unconditionally (without the if), wouldn't do much harm :)

Also, I generally prefer:

    for (const auto& path : m_path_history) {
    }

... for iteration. Avoids signedness warnings, too :)

File: Kernel/run [hunk 6]
```diff
@@ -38,6 +38,7 @@ else
         -device e1000,netdev=breh \
         -kernel kernel \
         -hda _fs_contents \
-        -soundhw pcspk
+        -soundhw pcspk \
+    	-debugcon stdio
```
Comment (awesomekling) [inline Kernel/run]: This change is unrelated and should not be in the PR.
Also, if you pull master, you'll find fba57d6ba3f0b5fc56d01874df2f37a0d21a32c3 that enabled this by default :)

File: Applications/FileManager/DirectoryView.cpp [hunk 7]
```diff
@@ -108,8 +110,32 @@ void DirectoryView::set_view_mode(ViewMode mode)
...
+            if (i == index)
+                dbgprintf("> ");
+            dbgprintf("Path: %s\n", path_history[i].view().characters());
+        }
+    }
+}
+
+void DirectoryView::add_path_to_history(const String& path)
+{
+    if (path_history_position < path_history.size())
```
Comment (rburchell) [inline Applications/FileManager/DirectoryView.cpp]: Vector::append does increase the capacity, so the resize here isn't necessary.

Small future note; perhaps it might be an idea to keep a cap on the size of the history? But no need to add that now, just a thought :)

**Comment by christopherdumas** [comment]
Ok, on it! I'll be back in a jiffy

File: Applications/FileManager/DirectoryView.cpp [hunk 8]
```diff
@@ -121,10 +147,29 @@ void DirectoryView::set_status_message(const String& message)
...
 }
 
 void DirectoryView::refresh()
 {
     model().update();
 }
+
+void DirectoryView::open_previous_directory()
+{
+    if (path_history_position > 0) {
```
Comment (rburchell) [inline Applications/FileManager/DirectoryView.cpp]: Small improvement opportunity: could consider disabling the actions when the path is at the start/end. I don't know if we have support for disabled actions, just something to think about :)

File: Kernel/run (hunk 6 above)
Comment (rburchell) [inline Kernel/run]: Not necessary (and it'll conflict, since kling just added it too :))

**Comment by rburchell** [comment]
Might also add a mention to the PR description that this fixes #36

**Comment by christopherdumas** [comment]
Okay I made the changes. I squashed the two commits into one, and it seems to have updated the PR correctly so I think we're all good! Before we merge though, let me do some digging and see if there's a way to disable actions.

**Comment by christopherdumas** [comment]
Okay. Sorry I explained the removal from the menu in the IRC. Basically, menus can't have disabled items, so having them in the menu caused a crash.

I'll rename those variables (:

**Comment by rburchell** [comment]
Summarising IRC discussion:

I suggested that removing them from the menu would be OK. Leaving them in hangs the client because WindowServer gets a request for an invalid menu that it doesn't respond to (and the client sends the request synchronously) -- a problem that should be fixed, but looks like a separate problem.

After looking at it, I think the underlying problem is actually that the menu isn't used at all right now (so WindowServer probably doesn't know about it.

@christopherdumas Try:

    menubar->add_menu(move(go_menu));

And see if that helps.

**Comment by christopherdumas** [comment]
Ok, awesome! I'll try that

**Comment by christopherdumas** [comment]
Yup, it works.

**Comment by christopherdumas** [comment]
Is there anything else I need to do before we merge this?

**Comment by awesomekling** [comment]
Looks cool, let’s merge it! Thank you for fixing this :)

**Comment by christopherdumas** [comment]
No problem!
